node/packages/sdk/lib/tags.js

```diff
@@ -0,0 +1,130 @@
+    const value = ensureString(inputValue, {
+      errorCode,
+      errorMessage: `Invalid trace span tag ${propertyName}: Expected string, received "%v"`,
+    });
+    if (isValidTagName(value)) return value;
+    return resolveException(inputValue, null, {
+      errorCode,
+      errorMessage:
+        `Invalid trace span tag ${propertyName}: ${capitalize.call(
+          propertyName
+        )} should contain dot separated tokens that follow ` +
+        '"[a-z][a-z0-9_]*" pattern. Received "%v"',
+    });
+  };
+})();
+
+const ensureTagValue = (() => {
+  const errorCode = 'INVALID_TRACE_SPAN_TAG_VALUE';
+  return (inputValue, tagName) => {
+    if (typeof inputValue === 'string') return inputValue;
+    if (typeof inputValue === 'number') {
+      return ensureFinite(inputValue, {
+        errorCode,
+        errorMessage:
+          `Invalid trace span tag value for "${tagName}": ` +
+          'Number must be finite, received "%v"',
+      });
+    }
+    if (typeof inputValue === 'boolean') return inputValue;
+    if (isDate(inputValue)) return inputValue.toISOString();
+    if (Array.isArray(inputValue)) {
+      let type = null;
+      return inputValue.map((item) => {
+        if (typeof item === 'string') {
+          if (type == null) type = 'string';
+          if (type === 'string') return item;
+        } else if (typeof item === 'number') {
+          if (type == null) type = 'number';
+          if (type === 'number') {
+            return ensureFinite(item, {
+              errorCode,
+              errorMessage:
+                `Invalid trace span tag value for "${tagName}": ` +
+                'Number must be finite, received "%v"',
+            });
+          }
+        } else if (isDate(inputValue)) {
+          if (type == null) type = 'date';
+          if (type === 'date') return inputValue.toISOString();
+        } else {
+          return resolveException(inputValue, null, {
+            errorCode,
+            errorMessage:
+              `Invalid trace span tag value for "${tagName}": ` +
+              'Unrecognized value type in array:"%v"',
+          });
+        }
+        return resolveException(inputValue, null, {
+          errorCode,
+          errorMessage:
+            `Invalid trace span tag value for "${tagName}": ` +
+            'Array cannot have mixed type values:"%v"',
+        });
+      });
+    }
+    return resolveException(inputValue, null, {
+      errorCode,
+      errorMessage:
+        `Invalid trace span tag value for "${tagName}": ` +
+        'Value must be either boolean, number, string, date or array of same values. Received "%v"',
+    });
+  };
+})();
+
+module.exports = class Tags extends Map {
+  set(inputName, value) {
+    const name = ensureTagName(inputName);
+    value = ensureTagValue(value, name);
+    if (this.has(name)) {
+      const currentValue = this.get(name);
+      if (Array.isArray(value)) {
+        if (Array.isArray(currentValue)) {
```

Comment:
Should we maybe throw a different error in this case? I know it is still technically a duplicate tag error but it is also a type mismatch if it throws when it fails the `Array.isArray(currentValue)` check here.

Reply:
This logic is taken as is from the current version of AWS Lambda SDK (it's not a new addition).

The idea is that we allow to set tag once and throw if there's an attempt to set the same tag again.
Still, to support some race condition scenarios where the same tag might be set with the same value by different parts of logic, I've added handling that, instead of throwing, ignores attempts of setting tag with the same value.

Now, I'm not sure what would be the extra value on producing different errors when the type and value is different. It definitely will increase the complexity of this check. 
Just throwing on different value feels good enough to me. We're addressing a rare scenario here, and the user is well-informed of what happened.